### PR TITLE
8334445: Parallel: Decouple maximum compaction from SoftReference clearing

### PR DESCRIPTION
**Pull request by albertnetymk**
Simple refactoring to not assume maximum-compaction from soft-ref-clearing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334445](https://bugs.openjdk.org/browse/JDK-8334445): Parallel: Decouple maximum compaction from SoftReference clearing (**Enhancement** - P4)

### Reviewers
 * [Zhengyu Gu](https://openjdk.org/census#zgu) (@zhengyu123 - **Reviewer**)
 * [Liang Mao](https://openjdk.org/census#lmao) (@mmyxym - no project role)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19763/head:pull/19763` \
`$ git checkout pull/19763`

Update a local copy of the PR: \
`$ git checkout pull/19763` \
`$ git pull https://git.openjdk.org/jdk.git pull/19763/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19763`

View PR using the GUI difftool: \
`$ git pr show -t 19763`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19763.diff">https://git.openjdk.org/jdk/pull/19763.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19763#issuecomment-2175574531)

### CODE REVIEW

File: src/hotspot/share/gc/parallel/psParallelCompact.cpp [hunk 1]
```diff
@@ -958,26 +963,23 @@ void PSParallelCompact::summary_phase(bool maximum_compaction)
...
-  ParallelScavengeHeap* heap = ParallelScavengeHeap::heap();
-  assert(!heap->is_stw_gc_active(), "not reentrant");
-
   IsSTWGCActiveMark mark;
 
-  const bool clear_all_soft_refs =
-    heap->soft_ref_policy()->should_clear_all_soft_refs();
+  ParallelScavengeHeap* heap = ParallelScavengeHeap::heap();
+  clear_all_soft_refs = clear_all_soft_refs
+                     || heap->soft_ref_policy()->should_clear_all_soft_refs();
```
Comment (zhengyu123) [inline src/hotspot/share/gc/parallel/psParallelCompact.cpp:975]: In `ParallelScavengeHeap::failed_mem_allocate()`, it does 

```do_full_collection(clear_all_soft_refs = false);
     ....
     do_full_collection(clear_all_soft_refs = true);
     ....
     throws OOM
```

There is possibility that `do_full_collection(clear_all_soft_refs = false)` cycle clears all soft refs due to soft ref policy. In this case, should we bypass the second `do_full_collection()` and throws OOM immediately?

Comment (albertnetymk) [reply src/hotspot/share/gc/parallel/psParallelCompact.cpp:975]: That seems a potential optimization. Can we defer it to a followup ticket? (I have some other cleanups in this area, which can affect the design decision here.)

Comment (zhengyu123) [reply src/hotspot/share/gc/parallel/psParallelCompact.cpp:975]: Sounds reasonable.

**Review by zhengyu123 — APPROVED** [review]
LGTM

File: src/hotspot/share/gc/parallel/psParallelCompact.hpp [hunk 2]
```diff
@@ -739,7 +738,7 @@ class PSParallelCompact : AllStatic {
   // make the heap parsable.
   static void fill_dense_prefix_end(SpaceId id);
 
-  static void summary_phase(bool maximum_compaction);
+  static void summary_phase();
```
[the remark this answers is not in the document]
Comment (mmyxym) [reply src/hotspot/share/gc/parallel/psParallelCompact.hpp:741]: I just thought the behavior change was significant. If it is expected, it looks good.

**Review by mmyxym — APPROVED** [review]

**Comment by albertnetymk** [comment]
Thanks for review.

/integrate

**Pull request closed without merging**